troubadix/standalone_plugins/allowed_rev_diff.py

```diff
@@ -0,0 +1,161 @@
+    argument_parser = ArgumentParser(
+        "Tool to check that the target rev only has diffs conforming to "
+        "the given patterns regarding the source rev"
+    )
+
+    argument_parser.add_argument(
+        "-d",
+        "--directory",
+        default=Path.cwd(),
+        type=Path,
+        help="The directory the repository to check is located in. "
+        "Defaults to 'pwd'",
+    )
+
+    ignored_linestart_group = argument_parser.add_mutually_exclusive_group()
+
+    ignored_linestart_group.add_argument(
+        "-i",
+        "--ignored-linestarts",
+        nargs="*",
+        type=str,
+        help="A list of line starts which will make the line be ignored. "
+        "Default: ['diff ', 'index ', '--- ', '+++ ', '@@ ']",
+    )
+
+    ignored_linestart_group.add_argument(
+        "--ignored-linestart-file",
+        type=Path,
+        help="The file containing a list of line starts, "
+        "separated by newlines, which will make the line be ignored. "
+        "Default: ['diff ', 'index ', '--- ', '+++ ', '@@ ']",
+    )
+
+    pattern_group = argument_parser.add_mutually_exclusive_group(required=True)
+
+    pattern_group.add_argument(
+        "-p",
+        "--patterns",
+        type=ensure_pattern,
+        nargs="*",
+        help="The list of patterns to check the diff for",
+    )
+
+    pattern_group.add_argument(
+        "--pattern-file",
+        type=Path,
+        help="The file containing the list of patterns, separated by newlines, "
+        "to check the diff for",
+    )
+
+    argument_parser.add_argument(
+        "-s", "--source", type=str, required=True, help="The upstream rev"
+    )
+
+    argument_parser.add_argument(
+        "-t", "--target", type=str, required=True, help="The downstream rev"
+    )
+
+    return argument_parser.parse_args()
+
+
+def check_diff_line_starts_with_ignored_linestart(
+    line: str, ignored_linestarts: List[str]
+) -> bool:
+    return any(
+        linestart
+        for linestart in ignored_linestarts
+        if line.startswith(linestart)
+    )
+
+
+def check_diff_line_matches_pattern(line: str, patterns: List[Pattern]) -> bool:
+    return any(pattern for pattern in patterns if pattern.match(line))
+
+
+def check_diff(
+    lines: List[str], ignored_linestarts: List[str], patterns: List[Pattern]
+) -> List[str]:
+    return [
+        line
+        for line in lines
+        if not check_diff_line_starts_with_ignored_linestart(
+            line, ignored_linestarts
+        )
+        and not check_diff_line_matches_pattern(line, patterns)
+    ]
+
+
+def read_ignored_linestarts(path: Path) -> List[str]:
+    if not path:
+        return None
+
+    with open(path, "r", encoding="UTF-8") as file:
+        return [line[:-1] for line in file.readlines()]
+
+
+def read_patterns(path: Path) -> List[Pattern]:
+    with open(path, "r", encoding="UTF-8") as file:
+        return [re.compile(pattern[:-1]) for pattern in file.readlines()]
```

Comment:
Is the `pattern[:-1]` to strip the line break? If so, using `.removesuffix("\n")` might be easier to understand

Reply:
True, fixed